### DISCUSSION TITLE
Jerky Slow Speed Fix!

### DISCUSSION THREAD

**Discussion by schmittlema** [prl-mushr/mushr#19] [Hardware]
If you have experienced jerky slow speed behavior with the MuSHR car like the video below then we have a recent fix for this problem.

https://user-images.githubusercontent.com/14075048/117983569-15fc5980-b305-11eb-949c-94f92b378213.mp4

### The fix
1. `cd ~/catkin_ws/src/mushr/mushr_base/vesc`
2. `git pull origin master`
3. `cd ~/catkin_ws`
4. `catkin_make`
5. Repeat steps 11 and 12 from [software setup](https://mushr.io/hardware/build_instructions/#software-setup)

### Cause of the Problem
There are a couple causes to this problem:
1. We were allowing erpm values to be sent to the vesc that were below the minimum but above 0. This causes some funky behavior, particularly paired with the fluctuations that happen from the joystick controller. Now, no values can fall into 0 - min_erpm
2. The PID integral gain in the internal vesc code was set too high. When giving a slow command, the car will seem smooth if not on the ground (because of fix 1). But once, placed on the ground the car has to move the weight of itself and thus apply more power to the motor. To adjust the power, the integral term in the PID loop increases the power untill the car starts moving, then slows as the car gets up to speed. Our integral term was too high, causing the car to lurch up to speed then rapidly drop power. We reduced the KI term by a factor of 10 and that seems to smooth out the behavior. If you experience problems with the current settings please comment and we can take a look.

**Comment by zlite** [comment]
Maybe update this to make clear where the vesc code should go. 

cd ~/catkin_ws/src/mushr/mushr_base/vesc
git pull https://github.com/prl-mushr/vesc.git
cd ~/catkin_ws
catkin_make

**Reply by schmittlema** [reply]
Thanks Chris for the suggestion! updated

**Comment by graj24** [comment]
Thanks for the solution. It worked for one of my cars but when I did the same thing to another car, it didn't work. The car still jerks while moving. When I ran the command "git pull origin master," it said that it's already up-to-date. I'm not sure what to do. Does it have anything to do with the wiring of vesc and brushless motor? There are 3 wires, is there a particular way to connect them?

**Reply by schmittlema** [reply]
Hi @graj24! If you are using a MuSHR car the vesc wiring instructions can be found [here](https://mushr.io/hardware/build_instructions/#vesc-preparation). Can you repeat steps 11 and 12 from [here](https://mushr.io/hardware/build_instructions/#software-setup) then go to "Advanced" in the BLDC tool and report back what the KI, KD, and KP terms are?
